x-pack/plugin/repository-encrypted/src/main/java/org/elasticsearch/repositories/encrypted/EncryptedRepository.java

```diff
@@ -3,14 +3,821 @@
+        }
+        return new String(Base64.getUrlEncoder().withoutPadding().encode(salt), StandardCharsets.UTF_8) + ":" +
+                new String(Base64.getUrlEncoder().withoutPadding().encode(hash), StandardCharsets.UTF_8);
+    }
+
+    /**
+     * Called before every snapshot operation on every node to validate that the snapshot metadata contains a password hash
+     * that matches up with the repository password on the local node.
+     *
+     * @param snapshotUserMetadata the snapshot metadata to verify
+     * @param exception the exception handler to call when the repository password check fails
+     */
+    private void validateRepositoryPasswordHash(Map<String, Object> snapshotUserMetadata, Consumer<Exception> exception) {
+        Object repositoryPasswordHash = snapshotUserMetadata.get(PASSWORD_HASH_RESERVED_USER_METADATA_KEY);
+        if (repositoryPasswordHash == null || (false == repositoryPasswordHash instanceof String)) {
+            exception.accept(new RepositoryException(metadata.name(), "Unexpected fatal internal error",
+                    new IllegalStateException("Snapshot metadata does not contain the repository password hash as a String")));
+            return;
+        }
+        if (false == passwordHashVerifier.verify((String) repositoryPasswordHash)) {
+            exception.accept(new RepositoryException(metadata.name(),
+                    "Repository password mismatch. The local node's value of the keystore secure setting [" +
+                            EncryptedRepositoryPlugin.ENCRYPTION_PASSWORD_SETTING.getConcreteSettingForNamespace(metadata.name()).getKey() +
+                            "] is different from the elected master node, which started the snapshot operation"));
+            return;
+        }
+    }
+
+    /**
+     * This is used to verify that salted hashes match up with the {@code password} from the constructor argument.
+     * This also caches the last successfully verified hash, so that repeated checks for the same hash turn into a simple {@code String
+     * #equals}.
+     */
+    private static class HashVerifier {
+        // the password to which the salted hashes must match up with
+        private final char[] password;
+        // the last successfully matched salted hash
+        private final AtomicReference<String> lastVerifiedHash;
+
+        HashVerifier(char[] password) {
+            this.password = password;
+            this.lastVerifiedHash = new AtomicReference<>(null);
+        }
+
+        boolean verify(String saltedHash) {
+            Objects.requireNonNull(saltedHash);
+            // first check if this exact hash has been checked before
+            if (saltedHash.equals(lastVerifiedHash.get())) {
+                return true;
+            }
+            String[] parts = saltedHash.split(":");
+            if (parts == null || parts.length != 2) {
```

Comment:
`parts` is never `null`. Also, shouldn't we throw on this or at least log the broken value here?

Reply:
Yes, this is the case of an internally broken state. I pushed 
a7e2dac , which logs it, but I'm still torn between which cases should throw, and which exception types. I must inspect (and have tests) for this failure cases. I'll get back to this problem.